File: ccpca/cpca.cpp
#include "cpca.hpp"

#include <Eigen/Eigenvalues>
#include <Eigen/QR>
#include <Eigen/SVD>
#include <chrono>
#include <cmath>
#include <iostream>
#include <utility>

CPCA::CPCA(Eigen::Index const nComponents, bool const standardize)
    : nComponents_(nComponents), standardize_(standardize) {
  initialize();
}

void CPCA::initialize() { components_.resize(0, 0); }

Eigen::MatrixXf
CPCA::fitTransform(Eigen::MatrixXf const &fg, Eigen::MatrixXf const &bg,
                   bool const autoAlphaSelection, float const alpha,
                   float const eta, float const convergenceRatio,
                   unsigned int maxIter, bool const keepReports) {
  fit(fg, bg, autoAlphaSelection, alpha, eta, convergenceRatio, maxIter,
      keepReports);
  return transform(fg_);
}

void CPCA::fit(Eigen::MatrixXf const &fg, Eigen::MatrixXf const &bg,
               bool const autoAlphaSelection, float const alpha,
               float const eta, float const convergenceRatio,
               unsigned int maxIter, bool const keepReports) {
  if (autoAlphaSelection) {
    fitWithBestAlpha(fg, bg, alpha, eta, convergenceRatio, maxIter,
                     keepReports);
  } else {
    fitWithManualAlpha(fg, bg, alpha);
  }
}

void CPCA::fitWithManualAlpha(Eigen::MatrixXf const &fg,
                              Eigen::MatrixXf const &bg, float const alpha) {
  fg_ = fg;
  bg_ = bg;

  Eigen::Index fgSize = fg_.size();
  Eigen::Index bgSize = bg_.size();

  if (fgSize == 0 && bgSize == 0) {
    std::cerr << "Both target and background matrices are empty." << std::endl;
  } else if (fgSize == 0) {
    // the result will be the same with when alpha is +inf
    fg_ = Eigen::MatrixXf::Zero(1, bg_.cols());
  } else if (bgSize == 0) {
    // the result will be the same with ordinary PCA
    bg_ = Eigen::MatrixXf::Zero(1, fg_.cols());
  }

  Eigen::Index nFeaturesFg = fg_.cols();
  Eigen::Index nFeaturesBg = bg_.cols();

  if (nFeaturesFg != nFeaturesBg) {
    std::cerr << "# of features of foregraound and background must be the same."
              << std::endl;
  }

  fg_ = fg_.rowwise() - fg_.colwise().mean();
  bg_ = bg_.rowwise() - bg_.colwise().mean();

  if (standardize_) {
    Eigen::RowVectorXf fgStd = fg_.array().square().colwise().mean().sqrt();
    Eigen::RowVectorXf bgStd = bg_.array().square().colwise().mean().sqrt();

    fg_ = fg_.array().rowwise() / fgStd.array();
    bg_ = bg_.array().rowwise() / bgStd.array();

    // NaN to 0.0f
    fg_ = fg_.unaryExpr([](float v) { return std::isfinite(v) ? v : 0.0f; });
    bg_ = bg_.unaryExpr([](float v) { return std::isfinite(v) ? v : 0.0f; });
  }

  fgCov_ = (fg_.adjoint() * fg_) /
           std::fmax(float(fg_.rows() - 1), std::numeric_limits<float>::min());
  bgCov_ = (bg_.adjoint() * bg_) /
           std::fmax(float(bg_.rows() - 1), std::numeric_limits<float>::min());

  Eigen::SelfAdjointEigenSolver<Eigen::MatrixXf> es(fgCov_ - alpha * bgCov_);
  components_ = es.eigenvectors().rightCols(nComponents_).rowwise().reverse();
  eigenvalues_ = es.eigenvalues().real().tail(nComponents_).reverse();
  totalPosEigenvalue_ = (es.eigenvalues().real().array() < 0)
                            .select(0, es.eigenvalues().real())
                            .sum();
  loadings_ = components_.array().rowwise() * eigenvalues_.array().abs().sqrt();
}

void CPCA::fitWithBestAlpha(Eigen::MatrixXf const &fg,
                            Eigen::MatrixXf const &bg, float const initAlpha,
                            float const eta, float const convergenceRatio,
                            unsigned int const maxIter,
                            bool const keepReports) {
  bestAlpha(fg, bg, initAlpha, eta, convergenceRatio, maxIter, keepReports);
  // updateComponents(bestAlpha_);
  fitWithManualAlpha(fg, bg, bestAlpha_);
}

void CPCA::updateComponents(float const alpha) {
  if (components_.cols() == 0) {
    std::cerr << "Run fit() at least once before updateComponents()"
              << std::endl;
  }
  Eigen::SelfAdjointEigenSolver<Eigen::MatrixXf> es(fgCov_ - alpha * bgCov_);
  components_ = es.eigenvectors().rightCols(nComponents_).rowwise().reverse();

  eigenvalues_ = es.eigenvalues().real().tail(nComponents_).reverse();
  totalPosEigenvalue_ = (es.eigenvalues().real().array() < 0)
                            .select(0, es.eigenvalues().real())
                            .sum();
  loadings_ = components_.array().rowwise() * eigenvalues_.array().abs().sqrt();
}

Eigen::MatrixXf CPCA::transform(Eigen::MatrixXf const &X) {
  if (components_.cols() == 0) {
    std::cerr << "Run fit() before transform()" << std::endl;
  }
  return X * components_;
}

float CPCA::bestAlpha(Eigen::MatrixXf const &fg, Eigen::MatrixXf const &bg,
                      float const initAlpha, float const eta,
                      float const convergenceRatio, unsigned int const maxIter,
                      bool const keepReports) {
  reports_.clear();
  float alpha = initAlpha;
  fit(fg, bg, false, alpha);

  // method 1. discard minor eigenvectors to avoid singular
  // Eigen::SelfAdjointEigenSolver<Eigen::MatrixXf> esQ(bgCov_);
  // float ratioToKeep = 0.999f;
  // Eigen::RowVectorXf eigenvalues = esQ.eigenvalues().real().reverse();
  // float targetTotalEigenVal = eigenvalues.sum() * ratioToKeep;
  // Eigen::Index nEigenVectorsToKeep = 0;
  // while (targetTotalEigenVal > 0) {
  //     targetTotalEigenVal -= eigenvalues(nEigenVectorsToKeep);
  //     nEigenVectorsToKeep++;
  // }
  // float ratioToKeep = 0.9f;
  // Eigen::Index nEigenVectorsToKeep = Eigen::Index(bg.cols() * ratioToKeep);
  //
  // std::cout << fg.cols() << " " << nEigenVectorsToKeep << std::endl;
  //
  // Eigen::MatrixXf Q =
  //     esQ.eigenvectors().rightCols(nEigenVectorsToKeep).rowwise().reverse();
  // Eigen::MatrixXf fgCovQ = Q.adjoint() * fgCov_ * Q;
  // Eigen::MatrixXf bgCovQ = Q.adjoint() * bgCov_ * Q;
  //
  // Eigen::SelfAdjointEigenSolver<Eigen::MatrixXf> esU(fgCovQ - alpha *
  // bgCovQ); Eigen::MatrixXf U =
  //     esU.eigenvectors().rightCols(nComponents_).rowwise().reverse();
  //
  // if (keepReports) {
  //   reports_.push_back(alpha);
  // }
  //
  // for (unsigned int i = 0; i < maxIter; ++i) {
  //   float fgTr = (U.adjoint() * fgCovQ * U).trace();
  //   float bgTr = (U.adjoint() * bgCovQ * U).trace();
  //   bgTr = std::fmax(bgTr, std::numeric_limits<float>::min());
  //   alpha = fgTr / bgTr;
  //
  //   // update U
  //   Eigen::SelfAdjointEigenSolver<Eigen::MatrixXf> esU(fgCovQ - alpha *
  //   bgCovQ); U =
  //   esU.eigenvectors().rightCols(nComponents_).rowwise().reverse();
  //
  //   if (keepReports) {
  //     reports_.push_back(alpha);
  //   }
  // }
  // bestAlpha_ = alpha;

  // method 2: add small constant to diag of bgCov_ to avoid singular
  bgCov_ += Eigen::MatrixXf::Identity(bgCov_.rows(), bgCov_.cols()) * eta;

  if (keepReports) {
    reports_.push_back(alpha);
  }

  for (unsigned int i = 0; i < maxIter; ++i) {
    float fgTr = (components_.adjoint() * fgCov_ * components_).trace();
    float bgTr = (components_.adjoint() * bgCov_ * components_).trace();
    bgTr = std::fmax(bgTr, std::numeric_limits<float>::min());

    float prevAlpha = alpha;
    alpha = fgTr / bgTr;
    updateComponents(alpha);
    if (keepReports) {
      reports_.push_back(alpha);
    }

    if (std::abs(prevAlpha - alpha) / alpha < convergenceRatio)
      break;
  }
  bestAlpha_ = alpha;

  return bestAlpha_;
}

std::vector<float> CPCA::logspace(float const start, float const end,
                                  unsigned int const num, float const base) {
  float realStart = std::pow(base, start);
  float realBase = std::pow(
      base, (end - start) /
                std::fmax(float(num - 1), std::numeric_limits<float>::min()));

  std::vector<float> result;
  result.reserve(num);
  std::generate_n(
      std::back_inserter(result), num, [=]() mutable throw()->float {
        float val = realStart;
        realStart *= realBase;
        return val;
      });
  return result;
}

// TODO: implement semi-automatic selection of alpha of the original cpca
// (but not necessary for ccPCA)

std::vector<float> CPCA::findSpectalAlphas(unsigned int const nAlphasToReturn,
                                           unsigned int const nAlphas,
                                           float const maxLogAlpha) {
  std::vector<float> alphas;
  Eigen::MatrixXf affinityMat = createAffinityMatrix(fg_, nAlphas, maxLogAlpha);

  // TODO: implement rest of here

  return alphas;
}

Eigen::MatrixXf CPCA::createAffinityMatrix(Eigen::MatrixXf const &X,
                                           unsigned int const nAlphas,
                                           float const maxLogAlpha) {
  std::vector<float> alphas;
  alphas.reserve(nAlphas + 1);
  alphas.push_back(0.0f);

  auto logspaceAlphas = logspace(-1.0f, maxLogAlpha, nAlphas);
  alphas.insert(alphas.end(), logspaceAlphas.begin(), logspaceAlphas.end());

  auto k = alphas.size();
  Eigen::MatrixXf affinityMat =
      0.5 * Eigen::MatrixXf::Identity(Eigen::Index(k), Eigen::Index(k));

  std::vector<Eigen::MatrixXf> subspaces;
  subspaces.reserve(k);
  for (auto const &alpha : alphas) {
    updateComponents(alpha);
    Eigen::MatrixXf proj = transform(X);
    Eigen::HouseholderQR<Eigen::MatrixXf> qr(proj);
    Eigen::MatrixXf Q = qr.householderQ();
    subspaces.push_back(qr.householderQ());
  }

  for (size_t i = 0; i < k; ++i) {
    for (size_t j = i + 1; j < k; ++j) {
      Eigen::BDCSVD<Eigen::MatrixXf> svd(subspaces[i] * subspaces[j],
                                         Eigen::ComputeThinU |
                                             Eigen::ComputeThinV);
      Eigen::VectorXf s = svd.singularValues();
      affinityMat(Eigen::Index(i), Eigen::Index(j)) = s(0) * s(1);
    }
  }

  affinityMat = affinityMat + affinityMat.transpose();
  // NaN to 0.0f
  affinityMat = affinityMat.unaryExpr(
      [](float v) { return std::isfinite(v) ? v : 0.0f; });

  return affinityMat;
}

File: cpca/cpca.py
import numpy as np
from scipy.linalg import eigh


class CPCA:
    """Contrastive PCA with efficient implemetation and automatic alpha selection.

    Parameters
    ----------
    n_components: int, optional, (default=2)
        A number of componentes to take.
    standardize: boo, optional, (default=True)
        Whether standardize input matrices or not.
    Attributes
    ----------
    None.
    Examples
    --------
    >>> import matplotlib.pyplot as plt
    >>> import numpy as np
    >>> from sklearn import datasets

    >>> from cpca import CPCA

    >>> dataset = datasets.load_iris()
    >>> X = dataset.data
    >>> y = dataset.target

    >>> # apply cPCA
    >>> cpca = CPCA()
    >>> X_new = cpca.fit_transform(fg=X, bg=X[y != 0], alpha=0.84)

    >>> # plot figures
    >>> plt.figure()
    >>> colors = ['navy', 'turquoise', 'darkorange']
    >>> lw = 2
    >>> for color, i, target_name in zip(colors, [0, 1, 2], [0, 1, 2]):
    ...     plt.scatter(
    ...         X_new[y == i, 0],
    ...         X_new[y == i, 1],
    ...         color=color,
    ...         alpha=.8,
    ...         lw=lw,
    ...         label=target_name)
    >>> plt.legend(loc='best', shadow=False, scatterpoints=1)
    >>> plt.title('cPCA of IRIS dataset with alpha=0.84')
    >>> plt.show()
    """

    def __init__(self, n_components=2, standardize=True):
        self.n_components = n_components
        self.standardize = standardize
        self.components_ = None
        self.eigenvalues_ = None
        self.total_pos_eigenvalue_ = None
        self.loadings_ = None
        self.best_alpha_ = None
        self.reports_ = None

        self._fg = None
        self._bg = None
        self._C_fg = None
        self._C_bg = None

    def fit(
        self,
        fg,
        bg,
        auto_alpha_selection=True,
        alpha=None,
        eta=1e-3,
        convergence_ratio=1e-2,
        max_iter=10,
        keep_reports=False,
    ):
        """Fit the model with a foreground matrix and a background matrix.

        Parameters
        ----------
        fg: array-like, shape (n_samples, n_features)
            A foreground (or target) dataset.
        bg: array-like, shape (n_samples, n_features)
            A background dataset. This column size must be the same size with
            fg. (A row size can be different from fg.)
        auto_alpha_selection:
            If True, find auto_alpha_selection for fit. Otherwise, compute PCs
            based on input alpha.
        eta: float, optional, (default=1e-3)
            Small constant value that will add to covariance matrix of bg when
            applying automatic alpha selection. Smaller eta tends to allow
            a larger alpha as the best alpha.
        convergence_ratio: float, optional, (default=1e-2)
            Threshold of improvement ratio for convergence of automatic alpha
            selection.
        max_iter=10: int, optional, (default=10)
            The number of alpha updates at most.
        keep_reports: bool, optional, (default=False)
            If True, while automatic alpha selection, reports are recorded. The
            reports are the history of "alpha" values.
        Returns
        -------
        self.
        """
        if alpha == None:
            alpha = 0.0
            auto_alpha_selection = True
        else:
            auto_alpha_selection = False

        if auto_alpha_selection:
            self._fit_with_best_alpha(
                fg, bg, alpha, eta, convergence_ratio, max_iter, keep_reports
            )
        else:
            self._fit_with_manual_alpha(fg, bg, alpha)

        return self

    def transform(self, X):
        """Obtaining transformed result Y with X and current PCs.

        Parameters
        ----------
        X : array-like, shape (n_samples, n_features)
            Testing data, where n_samples is the number of samples and
            n_features is the number of features. n_features must be the same
            size with the traiding data's features used for partial_fit.
        Returns
        -------
        Y : array-like, shape (n_samples, n_components)
            The transformed (or projected) result.
        """
        if self.components_ is None:
            print("Run fit() before transform()")
            return

        return X @ self.components_

    def fit_transform(
        self,
        fg,
        bg,
        auto_alpha_selection=True,
        alpha=None,
        eta=1e-3,
        convergence_ratio=1e-2,
        max_iter=10,
        keep_reports=False,
    ):
        """Fit the model with a foreground matrix and a background matrix and
        then obtain transformed result of fg and current PCs.

        Parameters
        ----------
        fg: array-like, shape (n_samples, n_features)
            A foreground (or target) dataset.
        bg: array-like, shape (n_samples, n_features)
            A background dataset. This column size must be the same size with
            fg. (A row size can be different from fg.)
        auto_alpha_selection:
            If True, find auto_alpha_selection for fit. Otherwise, compute PCs
            based on input alpha.
        alpha: float
            A contrast parameter, which quantifies the trade-off between having
            high target variance and low background variance. alpha must be
            equal to or larger than 0. If 0, the result will be the same with
            the ordinary PCA. If auto_alpha_selection is True, this alpha is
            used as an initial alpha value for auto selection.
        eta: float, optional, (default=1e-3)
            Small constant value that will add to covariance matrix of bg when
            applying automatic alpha selection. Smaller eta tends to allow
            a larger alpha as the best alpha.
        convergence_ratio: float, optional, (default=1e-2)
            Threshold of improvement ratio for convergence of automatic alpha
            selection.
        max_iter=10: int, optional, (default=10)
            The number of alpha updates at most.
        keep_reports: bool, optional, (default=False)
            If True, while automatic alpha selection, reports are recorded. The
            reports are the history of "alpha" values.
        Returns
        -------
        Y : array-like, shape (n_samples, n_components)
            The transformed (or projected) result.
        """
        self.fit(
            fg,
            bg,
            auto_alpha_selection,
            alpha,
            eta,
            convergence_ratio,
            max_iter,
            keep_reports,
        )
        return self.transform(fg)

    def update_components(self, alpha):
        """Update the components with a new contrast parameter alpha. Before
        using this, at least one time, fit() or fit_transform() must be called
        to build an initial result of the components.

        Parameters
        ----------
        alpha: float
            A contrast parameter, which quantifies the trade-off between having
            high target variance and low background variance. alpha must be
            equal to or larger than 0. If 0, the result will be the same with
            the ordinary PCA.
        Returns
        -------
        None.
        """
        if self.components_ is None:
            print("Run fit() at least once before update_components()")
            return
        self._update_components(alpha)

        return self

    # getter methods to be consistent with previous version of CPCA
    def get_components(self):
        """Returns current components.

        Parameters
        ----------
        None.
        Returns
        -------
        components: array-like, shape(n_features, n_components)
            Contrastive principal components.
        """
        return self.components_

    def get_component(self, index):
        """Returns i-th component.

        Parameters
        ----------
        index: int
            Indicates i-th component.
        Returns
        -------
        component: array-like, shape(1, n_components)
            i-th contrastive principal component.
        """
        if self.components_ is None:
            print("Run fit() before get_component()")
            return
        return self.components_[:, index]

    def get_eigenvalues(self):
        """Returns current eigenvalues.

        Parameters
        ----------
        None.
        Returns
        -------
        eigenvalues: array-like, shape(, n_components)
            Contrastive principal components' eigenvalues.
        """
        return self.eigenvalues_

    def get_eigenvalue(self, index):
        """Returns i-th eigenvalue.

        Parameters
        ----------
        index: int
            Indicates i-th eigenvalue.
        Returns
        -------
        eigenvalue: float
            i-th eigenvalue.
        """
        if self.eigenvalues_ is None:
            print("Run fit() before get_eigenvalue()")
            return
        return self.eigenvalues_[index]

    def get_total_pos_eigenvalue(self):
        """Returns the total of n_features positive eigenvalues (not n_components).
        This value can be used to compute the explained ratio of variance of the matrix C.

        Parameters
        ----------
        None
        Returns
        -------
        total_pos_eigenvalue: float
            The total of positive eigenvalues.
        """
        return self.total_pos_eigenvalue_

    def get_loadings(self):
        """Returns current principal component loadings.

        Parameters
        ----------
        None.
        Returns
        -------
        loadings: array-like, shape(n_features, n_components)
            Contrastive principal component loadings.
        """
        return self.loadings_

    def get_loading(self, index):
        """Returns i-th principal component loading.

        Parameters
        ----------
        index: int
            Indicates i-th principal component loading.
        Returns
        -------
        loading: array-like, shape(1, n_components)
            i-th principal component loading.
        """
        if self.loadings_ is None:
            print("Run fit() before get_loading()")
            return
        return self.loadings_[:, index]

    def get_reports(self):
        """Returns the reports kept while automatic selection of alpha. To get
        reports, you need to set keep_reports=True in the corresponding method.
        Parameters
        ----------
        None
        -------
        reports: array-like(n_alphas, 1),
            alpha values (at the same time optimization scores) obtained through
            best alpha selection".
        """
        return self.reports_

    def get_best_alpha(self):
        """Returns best alpha found with fit() with auto_selection_alpha=True
        Parameters
        ----------
        None
        -------
        best_alpha: float
            The found best alpha.
        """
        return self.best_alpha_

    def _update_components(self, alpha):
        w, v = eigh(self._C_fg - alpha * self._C_bg)
        self.components_ = v[:, np.argsort(-np.real(w))[: self.n_components]]
        self.eigenvalues_ = w[np.argsort(-np.real(w))[: self.n_components]]

        # Update related info
        self.total_pos_eigenvalue_ = self.eigenvalues_[self.eigenvalues_ > 0].sum()
        self.loadings_ = self.components_ * np.sqrt(np.abs(self.eigenvalues_))

        return self

    def _fit_with_manual_alpha(self, fg, bg, alpha):
        self._fg = fg
        self._bg = bg

        if self._fg.size == 0 and self._bg.size == 0:
            print("Both target and background matrices are empty.")
        elif self._fg.size == 0:
            # the result will be the same with when alpha is +inf
            self._fg = np.zeros((1, self._bg.shape[1]))
        elif self._bg.size == 0:
            # the result will be the same with ordinary PCA
            self._bg = np.zeros((1, self._fg.shape[1]))

        if self._fg.shape[1] != self._bg.shape[1]:
            print("# of features of foregraound and background must be the same.")

        self._fg -= np.mean(self._fg, axis=0)
        self._bg -= np.mean(self._bg, axis=0)

        if self.standardize:
            self._fg /= np.std(self._fg, axis=0)
            self._bg /= np.std(self._bg, axis=0)

            # NaN to 0.0f
            self._fg[np.isnan(self._fg)] = 0.0
            self._bg[np.isnan(self._bg)] = 0.0

        # Covariance matrix
        self._C_fg = (self._fg.T @ self._fg) / max(
            self._fg.shape[0] - 1, np.finfo(self._fg.dtype).tiny
        )
        self._C_bg = (self._bg.T @ self._bg) / max(
            self._bg.shape[0] - 1, np.finfo(self._bg.dtype).tiny
        )

        self._update_components(alpha)

        return self

    def _update_best_alpha(
        self, fg, bg, init_alpha, eta, convergence_ratio, max_iter, keep_reports
    ):
        """Automatic alpha selection for cPCA to maximize the variance ratio of target and backgroud.
        This implementation is based on:
        - Fujiwara et al., "Network Comparison with Interpretable Contrastive Network Representation Learning", JDSSV, 2022.
        """
        self.reports_ = []
        alpha = init_alpha
        self._fit_with_manual_alpha(fg, bg, alpha)

        # method 2: add small constant to diag of bgCov_ to avoid singular
        # (Note: refer to cpca.cpp to know what is method 1)
        self._C_bg += np.eye(self._C_bg.shape[0]) * eta

        if keep_reports:
            self.reports_.append(alpha)

        for _ in range(max_iter):
            tr_fg = np.trace(self.components_.T @ self._C_fg @ self.components_)
            tr_bg = np.trace(self.components_.T @ self._C_bg @ self.components_)
            tr_bg = max(tr_bg, np.finfo(self._C_bg.dtype).tiny)

            prev_alpha = alpha
            alpha = tr_fg / tr_bg
            self.update_components(alpha)
            if keep_reports:
                self.reports_.append(alpha)

            if (abs(prev_alpha - alpha) / alpha) < convergence_ratio:
                break

        self.best_alpha_ = alpha

        return self.best_alpha_

    def _fit_with_best_alpha(
        self, fg, bg, init_alpha, eta, convergence_ratio, max_iter, keep_reports
    ):
        self.best_alpha_ = self._update_best_alpha(
            fg, bg, init_alpha, eta, convergence_ratio, max_iter, keep_reports
        )
        self._fit_with_manual_alpha(fg, bg, self.best_alpha_)

        return self
